### PR TITLE
tests: latency_measure: reduce the chance of cycles underflow

### PR DESCRIPTION
**Pull request by ycsin**
Sometimes there's an unusually large cycles for tests that are known to complete with just a few cycles. Upon some testing, I found that it was because the overhead cycles was larger than the cycles taken by tests, causing the cycles to underflow.

To workaround that, make sure that the overhead measurement thread runs uninterrupted, repeat the measurement for a few times, and take the minimum value.

### CODE REVIEW

File: tests/benchmarks/latency_measure/src/timing_sc.c [hunk 1]
```diff
@@ -37,17 +37,28 @@ static void start_thread_entry(void *p1, void *p2, void *p3)
...
+	k_sched_lock();
+
+	for (int n = 0; n < 10; n++) {
+		start = timing_timestamp_get();
+		for (uint32_t i = 0; i < num_iterations; i++) {
+			timing_timestamp_get();
+		}
+		finish = timing_timestamp_get();
+
+		min_cycles = MIN(min_cycles, timing_cycles_get(&start, &finish));
```
Comment (ycsin) [inline tests/benchmarks/latency_measure/src/timing_sc.c:56]: The overhead measurement can vary quite a bit over the 10 iterations:
```
min_cycles: 7890
min_cycles: 6140
min_cycles: 6140
min_cycles: 3670
min_cycles: 2670
min_cycles: 2660
min_cycles: 2660
min_cycles: 2660
min_cycles: 2660
min_cycles: 2660
```

File: tests/benchmarks/latency_measure/src/timing_sc.c [hunk 2]
```diff
@@ -37,17 +37,28 @@ static void start_thread_entry(void *p1, void *p2, void *p3)
...
+	int key;
 
 	ARG_UNUSED(p2);
 	ARG_UNUSED(p3);
 
-	start = timing_timestamp_get();
-	for (uint32_t i = 0; i < num_iterations; i++) {
-		timing_timestamp_get();
+	key = irq_lock();
+	k_sched_lock();
```
Comment (peter-mitsis) [inline tests/benchmarks/latency_measure/src/timing_sc.c:47]: I like the idea of running this a few times to determine the minimum overhead value. However, locking interrupts and the scheduler using irq_lock() and k_sched_lock() respectively are not allowed from userspace. I note that some of the CI tests failed, and I suspect that this is the cause.

Admittedly I'm not certain what the source of the variance in the timings is--my initial suspicions are that a timer interrupt could be firing and getting serviced. If that is indeed the situation, then we can try to further reduce the likelihood of such an interruption by aligning the calibration to a tick boundary beforehand. The other item that comes to mind, is that some boards can not support a minimum system tick frequency of 1 Hz (they use an overlay to change it to a different value). I mention this as I have observed in the past that using a value too small for the timer driver to use can result in some messed up timing samples.

Comment (ycsin) [reply tests/benchmarks/latency_measure/src/timing_sc.c]: I've removed the locks, relying on repeated calculation of the overhead is good enough in my test

> aligning the calibration to a tick boundary beforehand.

How do we do that?

Comment (peter-mitsis) [reply tests/benchmarks/latency_measure/src/timing_sc.c]: In general, there are two ways to align to a tick boundary, or close enough to one.  The first (and simplest) is to issue a call to k_sleep(). When the thread wakes, assuming that no other thread is going to take precedence, the thread should be (nearly) aligned to a tick boundary.

```
k_sleep(K_TICKS(1));
```

The second is to a brute-force busy-wait approach. It too assumes that no other thread will take precedence in scheduling.

```
uint32_t current_tick = sys_clock_tick_get_32();

while (sys_clock_tick_get_32() == current_ticks) {
}
```

**Review by peter-mitsis — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**